staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go

```diff
@@ -20,6 +20,9 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// CustomResourceColumns masks the slice so protobuf can preserve empty slice
+type CustomResourceColumns []CustomResourceColumnDefinition
```

Comment:
remind me why we care about distinguishing nil and [] here?

Reply:
The concern was brought up in https://github.com/kubernetes/kubernetes/pull/67521#discussion_r215220972. My understanding is that we need to make protobuf preserve [], to allow overriding with an empty list. I didn't change the top-level field because it would be a breaking change.

Reply:
rereading that thread, it looks like the feedback there was to avoid making the distinction between nil and [] (to not do this nullable typedef). I think the `len(...) == 0` checks for emptiness are sufficient. Were there specific places where we needed to be able to tell nil from [] for some reason?

Reply:
Thanks for the explanation :) No the apiserver doesn't need to distinguish nil and [] (and defaulting happens for both cases).

Reply:
great, then let's drop this and just go with `[]CustomResourceColumnDefinition`